src/solvers/refinement/string_refinement.cpp

```diff
@@ -1320,111 +1374,151 @@ exprt fill_in_array_expr(const array_exprt &expr, std::size_t string_max_length)
-        "above"));
-    array_exprt &array_expr=to_array_expr(index_expr.array());
-
-    const typet &char_type = index_expr.array().type().subtype();
-
-    // Access to an empty array is undefined (non deterministic result)
-    if(array_expr.operands().empty())
-    {
-      expr = symbol_generator("out_of_bound_access", char_type);
-      return;
-    }
+static exprt substitute_array_access(
+  const if_exprt &if_expr,
+  const exprt &index,
+  const std::function<symbol_exprt(const irep_idt &, const typet &)>
+    &symbol_generator,
+  const bool left_propagate)
+{
+  // Substitute recursively in branches of conditional expressions
+  const exprt true_case = substitute_array_access(
+    index_exprt(if_expr.true_case(), index), symbol_generator, left_propagate);
+  const exprt false_case = substitute_array_access(
+    index_exprt(if_expr.false_case(), index), symbol_generator, left_propagate);
 
-    size_t last_index=array_expr.operands().size()-1;
+  return if_exprt(if_expr.cond(), true_case, false_case);
+}
 
-    exprt ite=array_expr.operands().back();
+static exprt substitute_array_access(
+  const index_exprt &index_expr,
+  const std::function<symbol_exprt(const irep_idt &, const typet &)>
+    &symbol_generator,
+  const bool left_propagate)
+{
+  const exprt &array = index_expr.array();
+
+  if(array.id() == ID_symbol)
+    return index_expr;
+  if(auto array_of = expr_try_dynamic_cast<array_of_exprt>(array))
+    return array_of->op();
+  if(auto array_with = expr_try_dynamic_cast<with_exprt>(array))
+    return substitute_array_access(
+      *array_with, index_expr.index(), left_propagate);
+  if(auto array_expr = expr_try_dynamic_cast<array_exprt>(array))
+    return substitute_array_access(
+      *array_expr, index_expr.index(), symbol_generator);
+  if(auto if_expr = expr_try_dynamic_cast<if_exprt>(array))
+    return substitute_array_access(
+      *if_expr, index_expr.index(), symbol_generator, left_propagate);
+
+  UNREACHABLE;
+}
 
-    if(ite.type()!=char_type)
+/// Create an equivalent expression where array accesses and 'with' expressions
+/// are replaced by 'if' expressions, in particular:
+///  * for an array access `arr[index]`, where:
+///    `arr := {12, 24, 48}` the constructed expression will be:
+///    `index==0 ? 12 : index==1 ? 24 : 48`
+///  * for an array access `arr[index]`, where:
+///    `arr := array_of(12) with {0:=24} with {2:=42}` the constructed
+///    expression will be: `index==0 ? 24 : index==2 ? 42 : 12`
+///  * for an array access `(g1?arr1:arr2)[x]` where `arr1 := {12}` and
+///    `arr2 := {34}`, the constructed expression will be: `g1 ? 12 : 34`
+///  * for an access in an empty array `{ }[x]` returns a fresh symbol, this
+///    corresponds to a non-deterministic result
+/// Note that if left_propagate is set to true, the `with` case will result in
+/// something like: `index <= 0 ? 24 : index <= 2 ? 42 : 12`
+/// \param expr: an expression containing array accesses
```

Comment:
What about "array access expression"?

Reply:
`expr` can be something else than an array access, but contain one. For instance `12 + 3 * c[i]` contains the access `c[i]` but is not itself an array access